src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -597,6 +597,145 @@ def update(self):
 
 
+class OMPTaskloopDirective(OMPDirective):
+    '''
+    Class representing an OpenMP TASKLOOP directive in the PSyIR.
+
+    :param list children: list of Nodes that are children of this Node.
+    :param parent: the Node in the AST that has this directive as a child.
+    :type parent: :py:class:`psyclone.psyir.nodes.Node`
+    :param grainsize: The grainsize value used to specify the grainsize \
+                      clause on this OpenMP directive. If this is None \
+                      the grainsize clause is not applied. Default \
+                      value is None.
+    :type grainsize: int or None.
+    :param num_tasks: The num_tasks value used to specify the num_tasks \
+                      clause on this OpenMP directive. If this is None \
+                      the num_tasks clause is not applied. Default value \
+                      is None.
+    :type num_tasks: int or None.
+    '''
+    def __init__(self, children=None, parent=None, grainsize=None,
+                 num_tasks=None):
+        self._grainsize = grainsize
+        self._num_tasks = num_tasks
+        super(OMPTaskloopDirective, self).__init__(children=children,
+                                                   parent=parent)
+
+    @property
+    def dag_name(self):
+        '''
+        :returns: the name to use in the DAG for this node.
+        :rtype: str
+        '''
+        _, position = self._find_position(self.ancestor(Routine))
+        return "OMP_taskloop_" + str(position)
+
+    def node_str(self, colour=True):
+        '''
+        Returns the name of this node with (optional) control codes
+        to generate coloured output in a terminal that supports it.
+
+        :param bool colour: whether or not to include colour control codes.
+
+        :returns: description of this node, possibly coloured.
+        :rtype: str
+        '''
+        return "{0}[OMP taskloop]".format(self.coloured_name(colour))
+
+    def validate_global_constraints(self):
+        '''
+        Perform validation checks that can only be done at code-generation
+        time.
+
+        :raises GenerationError: if this OMPTaskloopDirective is not \
+                                 enclosed within an OpenMP parallel \
+                                 region and an OpenMP single region.
+        :raises GenerationError: if this OMPTaskloopDirective has both \
+                                 a grainsize and num_tasks value \
+                                 specified.
+        '''
+        # It is only at the point of code generation that we can check for
+        # correctness (given that we don't mandate the order that a user
+        # can apply transformations to the code). As an orphaned taskloop
+        # directive, we must have an OMPParallelDirective as an ancestor
+        # somewhere back up the tree and an OMPSerialDirective as an
+        # ancestor back up the tree.
+        if not (self.ancestor(OMPParallelDirective,
```

Comment:
The OMPParalleDirective ancestor check will already be done by code-generating the ancestor OMPSerialDirective, right? I am not sure it worth being explicit and repeating the check here?

I guess the question is, if we already have a valid OMPSerialDirective, we always can add a Taskloop inside? Or there are other restrictions we should account for?

If there are no other restrictions I would remove the first check.

Reply:
I changed this, but I actually realise that the OMPSerialDirective is an abstract class, so it doesn't actually have any requirement to be inside an OMPParallelDirective (though all the implementations are). Is this still valid? @sergisiso

Reply:
I think its fine never mind.

Reply:
Both sub-classes have almost the same validate method just changing the name of the class in the error message. I think the right thing to do is to move the validate into the OMPSerialDirective and use self._text_name in the error message. Then it will be clear that is guaranteed for all subclasses.

Currently _text_name will need to be set for each sub-class (e.g. look at src/psyclone/psyir/nodes/call.py:59) but will be set automatically in the future.

Reply:
Ok, I've restructured the validate_global_constraints to be in the parent class.